code/modules/modular_computers/file_system/programs/research_app.dm

```diff
@@ -0,0 +1,218 @@
+		if (n.design_ids?.len)
+			node_cache[cid]["design_ids"] = list()
+			for (var/d in n.design_ids)
+				node_cache[cid]["design_ids"] += compress_id(d)
+		if (n.unlock_ids?.len)
+			node_cache[cid]["unlock_ids"] = list()
+			for (var/un in n.unlock_ids)
+				node_cache[cid]["unlock_ids"] += compress_id(un)
+		if (n.required_experiments?.len)
+			node_cache[cid]["required_experiments"] = n.required_experiments
+		if (n.discount_experiments?.len)
+			node_cache[cid]["discount_experiments"] = n.discount_experiments
+
+	// Build design cache
+	var/design_cache = list()
+	var/datum/asset/spritesheet/research_designs/ss = get_asset_datum(/datum/asset/spritesheet/research_designs)
+	var/size32x32 = "[ss.name]32x32"
+	for (var/did in SSresearch.techweb_designs)
+		var/datum/design/d = SSresearch.techweb_designs[did] || SSresearch.error_design
+		var/cid = "[compress_id(d.id)]"
+		var/size = ss.icon_size_id(d.id)
+		design_cache[cid] = list(
+			d.name,
+			"[size == size32x32 ? "" : "[size] "][d.id]"
+		)
+
+	// Ensure id cache is included for decompression
+	var/flat_id_cache = list()
+	for (var/id in id_cache)
+		flat_id_cache += id
+
+	.["static_data"] = list(
+		"node_cache" = node_cache,
+		"design_cache" = design_cache,
+		"id_cache" = flat_id_cache
+	)
+
+/**
+ * Compresses an ID to an integer representation using the id_cache, used for deduplication
+ * in sent JSON payloads
+ *
+ * Arguments:
+ * * id - the ID to compress
+ */
+/datum/computer_file/program/science/proc/compress_id(id)
+	if (!id_cache[id])
+		id_cache[id] = id_cache_seq++
+	return id_cache[id]
+
+/datum/computer_file/program/science/proc/research_node(id, mob/user)
+	if(!stored_research.available_nodes[id] || stored_research.researched_nodes[id])
+		computer.say("Node unlock failed: Either already researched or not available!")
+		return FALSE
+	var/datum/techweb_node/TN = SSresearch.techweb_node_by_id(id)
+	if(!istype(TN))
+		computer.say("Node unlock failed: Unknown error.")
+		return FALSE
+	var/list/price = TN.get_price(stored_research)
+	if(stored_research.can_afford(price))
+		computer.investigate_log("[key_name(user)] researched [id]([json_encode(price)]) on techweb id [stored_research.id] via [computer].", INVESTIGATE_RESEARCH)
+		if(stored_research == SSresearch.science_tech)
+			SSblackbox.record_feedback("associative", "science_techweb_unlock", 1, list("id" = "[id]", "name" = TN.display_name, "price" = "[json_encode(price)]", "time" = SQLtime()))
+		if(stored_research.research_node_id(id))
+			computer.say("Successfully researched [TN.display_name].")
+			var/logname = "Unknown"
+			if(isAI(user))
+				logname = "AI: [user.name]"
+			if(iscarbon(user))
+				var/obj/item/card/id/idcard = user.get_active_held_item()
+				if(istype(idcard))
+					logname = "User: [idcard.registered_name]"
+			if(ishuman(user))
+				var/mob/living/carbon/human/H = user
+				var/obj/item/I = H.wear_id
+				if(istype(I))
+					var/obj/item/card/id/ID = I.GetID()
```

Comment:
```suggestion
				var/obj/item/worn_item = H.wear_id
				if(istype(worn_item))
					var/obj/item/card/id/ID = worn_item.GetID()
```
This line does not meet our contribution standards, specifically for [variable name compliance](https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names), and should be refactored.